packages/hasher/src/workspaces/fastFindUp.ts

```diff
@@ -0,0 +1,71 @@
+import path from "path";
+import fs from "fs-extra";
+
+interface FoundRecord {
+  cwd: string;
+  target: string;
+}
+
+let findUpCache = new Map<string, FoundRecord | undefined>();
+let root: string;
+
+function checkFoundRecords(targets: string[], cwd: string) {
+  for (const found of findUpCache.values()) {
+    if (found) {
+      for (const target of targets) {
+        if (cwd.startsWith(found.cwd) && found.target === target) {
+          return path.join(found.cwd, found.target);
+        }
+      }
+    }
+  }
+}
+
+export async function fastFindUp(targets: string[] | string, cwd: string) {
+  if (!root) {
+    root = path.parse(cwd).root;
+  }
+
+  targets = Array.isArray(targets) ? targets : [targets];
+
+  // check to see if this cwd is inside one of the found records
+  const fastFound = checkFoundRecords(targets, cwd);
+  if (fastFound) {
+    return fastFound;
+  }
+
+  const traversed: string[] = [];
+  let found: FoundRecord | undefined;
+  let needle: string;
+
+  done: {
+    while (cwd !== root) {
+      for (const target of targets) {
+        needle = path.join(cwd, target);
+
+        if (findUpCache.has(needle)) {
+          found = findUpCache.get(needle);
+          break done;
+        } else if (fs.existsSync(needle)) {
+          traversed.push(needle);
+          found = { cwd, target };
+          break done;
+        } else {
+          traversed.push(needle);
+        }
+      }
+
+      cwd = path.dirname(cwd);
+    }
+  }
+
+  for (const key of traversed) {
+    findUpCache.set(key, found);
+  }
+
+  if (found) {
+    return path.join(found.cwd, found.target);
+  } else {
+    return null;
+  }
+}
```

Comment:
Could we use `mem` and `find-up` together? Or would we miss some important optimizations?

```suggestion
import * as findUp from "find-up";
import * as mem from "mem";

export const fastFindUp = mem(findUp, { cacheKey: JSON.stringify });
```

Reply:
Here's the motivation behind the `cacheKey` strategy: https://github.com/sindresorhus/mem#example-options-like-argument

Reply:
So I'm not completely satisfied with what I have here. Maybe mem will solve it elegantly. I want to make sure we don't draw up the tree more than is needed. And, my optimization is also that for a list of KNOWN workspaces, use simple string match to look at whether the file belongs in that workspace. This string match is way faster than even having to look up the file system even one file existence.

Reply:
I see your point now. I did some experimentation comparing your solution with mem+fastUp. I counted cache hits + cache misses. In order to get the identical behavior, you could do something like:

```
import findUp from "find-up";
import mem from "mem";

export const fastFindUp: typeof findUp = mem(findUp, {
  cacheKey: args => (args[0] === ".git" ? ".git" : JSON.stringify(args))
});
```

It's adding a nasty implementation detail with the specific `.git`. Typically you'd want the cache key to be file + cwd. But maybe in our case we expect all of the things we do `findUp` on (.git, workspace and lock files) to only have one version. In which case we could simply do:

```
import findUp from "find-up";
import mem from "mem";

export const fastFindUp: typeof findUp = mem(findUp);
```

By default, mem will only use the first argument as the cache key.

If you want to do something like this, we need to convert the second argument of fastFindUp to an object, and specify that `.git` is a folder:

```
let gitRoot = await fastFindUp(".git", {
    cwd: packageRoot,
    type: "directory"
  });
```

Reply:
So, if we do this, every single package in a monorepo will at least get a file existsSync check (which apparently does take up a bunch of time - O(n)). mem(findup) will prevent O(n * m) where m is the depth of the packages. What the checkFoundRecords() thing does is a string based match that prevents the file existence checks. I'm gonna see if I can prevent the "findup"... so it keeps a list of recently found things:

1. if found - keep it around; all further find-ups will use a normalized string match instead of actually looking for a file
2. if not found in recently hit list (e.g. going from looking for workspace file to looking for a lockfile), it would then benefit from having a memoizing findup so that it will at most traverse once on the paths up

So, I think I can replace (2) with that `mem()` function. I still have to figure out maybe if (1) needs to be handled at a layer higher (like better memoization in the lockfile / getworkspaces files)

Reply:
Actually thinking about it some more there are some issues with `mem()` and `find-up` because unless find-up is using that memoized version of find-up internally, it'll continue to be memoization MISSes because we keep asking for find-up with different cwd values per package.

In thinking this through more, I think we might want to change the API of generateHash function to take in a workspaces param. If it is undefined, then we can feel free to do the old behavior... I am gonna attempt this so backfill + api usage can be fast. Will work on this next.

Reply:
Okay, I think we had a good chat, I'm avoiding doing find-up within this implementation. We moved that stuff into `workspace-tools`. What I am doing now is re-using repo info as much as possible.